Tests/PackageLoading/InMemoryConventionTests.swift

```diff
@@ -0,0 +1,174 @@
+import XCTest
+
+import Basic
+import PackageDescription
+import PackageLoading
+import PackageModel
+
+// FIXME: These test Utilities can/should be moved to test-specific library when we start supporting them.
+private extension FileSystem {
+    /// Create a file on the filesystem while recursively creating the parent directory tree.
+    ///
+    /// - Parameters:
+    ///     - file: Path of the file to create.
+    ///     - contents: Contents of the file. Empty by default.
+    ///
+    /// - Throws: FileSystemError
+    mutating func create(_ file: AbsolutePath, contents: ByteString = ByteString()) throws {
+        // Auto create the tree.
+        try createDirectory(file.parentDirectory, recursive: true)
+        try writeFileContents(file, bytes: contents)
+    }
+
+    /// Create multiple empty files on the filesystem while recursively creating the parent directory tree.
+    ///
+    /// - Parameters:
+    ///     - files: Paths of empty files to create.
+    ///
+    /// - Throws: FileSystemError
+    mutating func createEmptyFiles(_ files: AbsolutePath...) throws {
+        // Auto create the tree.
+        for file in files {
+            try createDirectory(file.parentDirectory, recursive: true)
+            try writeFileContents(file, bytes: ByteString())
+        }
+    }
+}
+
+/// Loads a package using PackageBuilder at the given path. If package description is provided it is used,
+/// otherwise a default package with name of directory is passed to manifest.
+///
+/// - Parameters:
+///     - path: Directory where the package is located.
+///     - in: FileSystem in which the package should be loaded from.
+///     - package: PackageDescription instance to use for loading this package.
+///
+/// - Throws: ModuleError, ProductError
+private func loadPackage(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil) throws -> PackageModel.Package {
+    let pkg: PackageDescription.Package
+    if let package = package {
+        pkg = package
+    } else {
+        pkg = Package(name: path.basename)
+    }
+    let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: pkg, products: [], version: nil)
+    let builder = PackageBuilder(manifest: manifest, path: path, fileSystem: fs)
+    return try builder.construct(includingTestModules: true)
+}
+
+extension PackageBuilderTester {
+    @discardableResult
+    init(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil, result: (PackageBuilderTester) -> Void) {
+        do {
+            self = .package(try loadPackage(path, in: fs, package: package))
+        } catch {
+            self = .error(error)
+        }
+        result(self)
```

Comment:
I would love it if this would check for a couple of things automatically:

1. Check that all errors or warnings were checked by the result body, or if the test wants to ignore them it should have to call a function (`ignoreDiagnostics`, say). This will force us to maintain our tests of diagnostics, which is good.

2. Check that all modules were checked, unless the test calls some function like `ignoreOtherModules`.

We just need to keep a list of the things which have been checked, and validate they are empty at the end.